stl/inc/xutility

```diff
@@ -4187,6 +4187,313 @@ struct default_sentinel_t {};
+        _STL_VERIFY(_Diff <= _Length, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return counted_iterator(_Current + _Diff, _Length - _Diff);
+    }
+
+    friend constexpr counted_iterator operator+(iter_difference_t<_Iter> _Diff, const counted_iterator& _Right)
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Right.count(), "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _Right + _Diff;
+    }
+
+    constexpr counted_iterator& operator+=(iter_difference_t<_Iter> _Diff)
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Length, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        _Current += _Diff;
+        _Length -= _Diff;
+        return *this;
+    }
+
+    constexpr counted_iterator operator-(iter_difference_t<_Iter> _Diff) const
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(-_Diff <= _Length, "counted_iterator decrement beyond end of range");
+        if constexpr (_Offset_verifiable_v<_Iter>) {
+            _Current._Verify_offset(-_Diff);
+        }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return counted_iterator(_Current - _Diff, _Length + _Diff);
+    }
+    // clang-format on
+
+    template <common_with<_Iter> _Other>
+    friend constexpr iter_difference_t<_Other> operator-(
+        const counted_iterator& _Left, const counted_iterator<_Other>& _Right) {
+        _Verify_Range(_Left, _Right);
+        return _Right._Length - _Left._Length;
+    }
+    friend constexpr iter_difference_t<_Iter> operator-(const counted_iterator& _Left, default_sentinel_t) {
+        return -_Left._Length;
+    }
+    friend constexpr iter_difference_t<_Iter> operator-(default_sentinel_t, const counted_iterator& _Right) {
+        return _Right._Length;
+    }
+
+    // clang-format off
+    constexpr counted_iterator& operator-=(iter_difference_t<_Iter> _Diff)
+        requires random_access_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(-_Diff <= _Length, "counted_iterator decrement beyond end of range");
+        if constexpr (_Offset_verifiable_v<_Iter>) {
+            _Current._Verify_offset(-_Diff);
+        }
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        _Current -= _Diff;
+        _Length += _Diff;
+        return *this;
+    }
+    // clang-format on
+
+    // [counted.iter.cmp]
+    template <common_with<_Iter> _Other>
+    friend constexpr bool operator==(const counted_iterator& _Left, const counted_iterator<_Other>& _Right) {
+        _Verify_Range(_Left, _Right);
+        return _Left._Length == _Right._Length;
+    }
+
+    friend constexpr bool operator==(const counted_iterator& _Left, default_sentinel_t) noexcept // strengthened

```

Comment:
We use `/* strengthened */` on non-constructors to avoid interfering with the brace flow.

Reply:
Ok, although the result is ugly